geometric_kernels/spaces/special_orthogonal_group.py

```diff
@@ -0,0 +1,408 @@
+                + B.cast(B.dtype_float(self.rank), self.rank)
+                - B.range(
+                    B.dtype_float(sgn),
+                    B.shape(sgn)[-1],
+                )
+                - 1 / 2
+            )  # [M, R]
+            ret = self.xi1(qs, gamma) / self.xi1(
+                B.range(self.rank, 0, -1)[None, :] - 1 / 2, gamma
+            )  # [..., M]
+            return ret
+        else:
+            qs = (
+                B.cast(B.dtype_float(sgn), sgn[:, :-1])
+                + B.cast(B.dtype_float(self.rank), self.rank)
+                - B.range(B.dtype_float(sgn), B.shape(sgn)[-1] - 1)
+                - 1
+            )  # [M, R - 1]
+            qs = B.concat(
+                qs, B.cast(B.dtype(qs), B.abs(sgn[:, None, self.rank - 1])), axis=1
+            )  # [M, R]
+            ret = B.where(
+                sgn[:, -1] == 0,
+                self.xi0(qs, gamma)
+                / self.xi0(B.range(B.dtype(qs), self.rank - 1, -1, -1)[None, :], gamma),
+                (
+                    self.xi0(qs, gamma)
+                    + self.xi1(qs, gamma) * B.cast(B.dtype(gamma), B.sign(sgn[:, -1]))
+                )
+                / self.xi0(B.range(B.dtype(qs), self.rank - 1, -1, -1)[None, :], gamma),
+            )
+            return ret
+
+    def _generate_signatures(self, num_repr):
+        r"""
+        Generate signatures and dimensions of representations and the corresponding eigenvalues.
+        """
+        signatures = []
+        for signature_sum in range(0, num_repr):
+            for i in range(1, self.rank + 1):
+                for signature in fixed_length_partitions(signature_sum, i):
+                    signature.extend([0] * (self.rank - i))
+                    signatures.append(tuple(signature))
+                    if self.dim % 2 == 0 and signature[-1] != 0:
+                        signature[-1] = -signature[-1]
+                        signatures.append(tuple(signature))
+
+        def _compute_dim(signature):
+            if self.dim % 2 == 1:
+                qs = [pk + self.rank - k - 1 / 2 for k, pk in enumerate(signature)]
+                rep_dim = reduce(
+                    operator.mul,
+                    (
+                        2 * qs[k] / math.factorial(2 * k + 1)
+                        for k in range(0, self.rank)
+                    ),
+                ) * reduce(
+                    operator.mul,
+                    (
+                        (qs[i] - qs[j]) * (qs[i] + qs[j])
+                        for i, j in it.combinations(range(self.rank), 2)
+                    ),
+                    1,
+                )
+                return int(round(rep_dim))
+            else:
+                qs = [
+                    pk + self.rank - k - 1 if k != self.rank - 1 else abs(pk)
+                    for k, pk in enumerate(signature)
+                ]
+                rep_dim = int(
+                    reduce(
```

Comment:
Is this going to be performant?

Reply:
Or would it be faster to use LAB functions to do this?

Reply:
It does not take long to compute (for a reasonable number of eigfs). Could be possible to rewrite it in numpy (LAB per se does not induce any performance gains) but I don't know if it worth it